engine/apiserver.go

```diff
@@ -0,0 +1,906 @@
+package engine
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/pprof"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/common/crypto"
+	"github.com/thrasher-corp/gocryptotrader/config"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/database/repository/exchange"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+// setupAPIServerManager checks and creates an api server manager
+func setupAPIServerManager(remoteConfig *config.RemoteControlConfig, pprofConfig *config.Profiler, exchangeManager iExchangeManager, bot iBot, portfolioManager iPortfolioManager, configPath string) (*apiServerManager, error) {
+	if remoteConfig == nil {
+		return nil, errNilRemoteConfig
+	}
+	if pprofConfig == nil {
+		return nil, errNilPProfConfig
+	}
+	if exchangeManager == nil {
+		return nil, errNilExchangeManager
+	}
+	if bot == nil {
+		return nil, errNilBot
+	}
+	if configPath == "" {
+		return nil, errEmptyConfigPath
+	}
+	return &apiServerManager{
+		remoteConfig:           remoteConfig,
+		pprofConfig:            pprofConfig,
+		restListenAddress:      remoteConfig.DeprecatedRPC.ListenAddress,
+		websocketListenAddress: remoteConfig.WebsocketRPC.ListenAddress,
+		exchangeManager:        exchangeManager,
+		bot:                    bot,
+		gctConfigPath:          configPath,
+		portfolioManager:       portfolioManager,
+	}, nil
+}
+
+// IsRunning safely checks whether the subsystem is running
+func (m *apiServerManager) IsRunning() bool {
+	if m == nil {
+		return false
+	}
+	return atomic.LoadInt32(&m.started) == 1
+}
+
+// Stop attempts to shutdown the subsystem
+func (m *apiServerManager) Stop() error {
+	if m == nil {
+		return fmt.Errorf("api server %w", ErrNilSubsystem)
+	}
+	if !atomic.CompareAndSwapInt32(&m.started, 1, 0) {
+		return fmt.Errorf("api server %w", ErrSubSystemNotStarted)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Nanosecond)
```

Comment:
😄 Yeah it might be, however as its a deprecated service used to communicate with the website, when shutting down I felt I could be forceful. Happy to increase it, though I don't think there are any consequences for killing it harshly. I'll double check